Add cycle instances task

sevenmachines opened the pull request:
THis change adds a task to cycle the instances on an autoscaling
group, plus additional supporting methods for the Autoscale
class. This functionality means that we can recreate instances
sequentially in an autoscaling group, useful when we want to ensure
we are using the latest AMI's with the newest security fixes.

Commit:
Add cycle instances task

THis change adds a task to cycle the instances on an autoscaling
group, plus additional supporting methods for the Autoscale
class. This functionality means that we can recreate instances
sequentially in an autoscaling group.

sevenmachines commented:
@bennythejudge WIP task on cycling instances, seems to work ok on testing. If you fancy using this as the base of our security task I think it might be the way to go.

sevenmachines commented:
We need to add a test to this to make sure that at least one healthcheck is run on the instance before we pass it as healthy. The ASG set the new instance to healthy while its in its 'grace' period and only checks the health after that time has expired. Solutions could be,
- (best) Check that the ASG has run a healthcheck on the new instance before judging it 'healthy'
- (works-but-not-great) Wait until after the grace period before checking instance health

Commit:
Modified instance-cycling
  - Added a delay to wait the ASG::HealthCheckGracePeriod
  - Add missing VPC class import to fab_tasks
  - Removed a `pass` statement no longer needed
  - Added a check on number of instances during the instance cycling loop

Commit:
Merge branch 'add_cycle_instances_task' into check_asg_health_check_is_done

Commit:
Merge pull request #188 from ministryofjustice/check_asg_health_check_is_done

Check asg health check is done

Commit:
Change placement of instance termination log message

The log message needs to be right before the actual termination
to make more sense.

Commit:
Add simple countdown for long waits

When we're waiting long periods for events, such as in the case of
an autoscaling event, this change will add the disaply of a simple
countdown so we don't look unresponsive.

filipposc5 committed:
Merge pull request #189 from ministryofjustice/add_simple_countdown_for_long_waits

Add simple countdown for long waits

filipposc5 commented on bootstrap_cfn/autoscale.py:
```diff
@@ -46,3 +53,145 @@ def get_all_autoscaling_groups(self):
+        current_instance_ids = [instance.get('InstanceId') for instance in self.get_healthy_instances()]
+        logging.getLogger("bootstrap-cfn").info("cycle_instances: Found {} instance ids, {}"
+                                                .format(len(current_instance_ids), current_instance_ids))
+
```
Is it worth presenting a status of all instances at this stage (not just the healthy ones)? Assuming someone will be changing their states it might be worth giving out a quick check like: 
`Healthy: 3  Unhealthy: 0`  <-- this would at least make you feel better that there were 0 unhealthy
Or something like `Healthy:  i-whatever1, i-whatever2    Other: i-whatever3`  <-- even better, it shows you which was not healthy/terminating/changing to begin with.

This would help us track "what happened" in case for example one of the instances was not healthy to begin with.

filipposc5 commented on bootstrap_cfn/autoscale.py:
```diff
@@ -46,3 +53,145 @@ def get_all_autoscaling_groups(self):
+    def set_autoscaling_desired_capacity(self,
+                                         capacity):
+        """
+        Set the desired instances count on an autosaling group
```
"autoscaling"

Commit:
Increase logged details about instances

This change means that we'll get a print out of all instances
at some points of the process so we can track changes and
any problems with the scaling process.

* Fix autosaling typo
* Add all instances display to initial logging list of current instances

sevenmachines commented:
note: needs squishing

sevenmachines commented:
Did final squish
